tests/expose_test.go

```diff
@@ -0,0 +1,384 @@
+				By("Getting back the cluster IP given for the service")
+				svc, err := virtClient.CoreV1().Services(vm.Namespace).Get(serviceName, k8smetav1.GetOptions{})
+				Expect(err).ToNot(HaveOccurred())
+				serviceIP := svc.Spec.ClusterIP
+
+				By("Starting a pod which tries to reach the VM via ClusterIP")
+				job := tests.NewHelloWorldJob(serviceIP, servicePort)
+				job, err = virtClient.CoreV1().Pods(vm.Namespace).Create(job)
+				Expect(err).ToNot(HaveOccurred())
+
+				By("Waiting for the pod to report a successful connection attempt")
+				getStatus := func() k8sv1.PodPhase {
+					pod, err := virtClient.CoreV1().Pods(job.Namespace).Get(job.Name, k8smetav1.GetOptions{})
+					Expect(err).ToNot(HaveOccurred())
+					return pod.Status.Phase
+				}
+				Eventually(getStatus, 60, 1).Should(Equal(k8sv1.PodSucceeded))
+			})
+		})
+
+		Context("Expose NodePort service", func() {
+			const servicePort = "27017"
+			const serviceName = "node-port-vm"
+			const nodePort = "30017"
+
+			It("Should expose a NodePort service on a VM", func() {
+				virtctl := tests.NewRepeatableVirtctlCommand(expose.COMMAND_EXPOSE, "virtualmachine", "--namespace",
+					vm.Namespace, vm.Name, "--port", servicePort, "--name", serviceName, "--target-port", strconv.Itoa(testPort),
+					"--type", "NodePort", "--node-port", nodePort)
+				err := virtctl()
+				Expect(err).ToNot(HaveOccurred())
+			})
+
+			It("Should be able to connect to the exposed service", func() {
+				By("Getting back the the service")
+				_, err := virtClient.CoreV1().Services(vm.Namespace).Get(serviceName, k8smetav1.GetOptions{})
+				Expect(err).ToNot(HaveOccurred())
+
+				By("Getting the node IP from all nodes")
+				nodes, err := virtClient.CoreV1().Nodes().List(k8smetav1.ListOptions{})
+				Expect(err).ToNot(HaveOccurred())
+				Expect(nodes.Items).ToNot(BeEmpty())
+				for _, node := range nodes.Items {
+					Expect(node.Status.Addresses).ToNot(BeEmpty())
+					nodeIP := node.Status.Addresses[0].Address
+
+					By("Starting a pod which tries to reach the VM via NodePort")
+					job := tests.NewHelloWorldJob(nodeIP, nodePort)
+					job, err = virtClient.CoreV1().Pods(vm.Namespace).Create(job)
+					Expect(err).ToNot(HaveOccurred())
+
+					By("Waiting for the pod to report a successful connection attempt")
+					getStatus := func() k8sv1.PodPhase {
+						pod, err := virtClient.CoreV1().Pods(job.Namespace).Get(job.Name, k8smetav1.GetOptions{})
+						Expect(err).ToNot(HaveOccurred())
+						return pod.Status.Phase
+					}
+					Eventually(getStatus, 60, 1).Should(Equal(k8sv1.PodSucceeded))
+				}
+			})
+		})
+	})
+
+	Context("Expose UDP service on a VM", func() {
+		By("Creating a VM object")
+		vm := tests.NewRandomVMWithEphemeralDiskAndUserdata(tests.RegistryDiskFor(tests.RegistryDiskCirros), "#!/bin/bash\necho 'hello'\n")
+		vm.Labels = map[string]string{"expose": "udp-vm"}
+
+		It("Should start a VM and run a 'hello world' UDP server on it", func() {
```

Comment:
Same like above make a `BeforeAll` out of this.

Reply:
will fix